device/core/src/sak_authentication_provider.ts

```diff
@@ -104,27 +98,41 @@ export class SharedAccessKeyAuthenticationProvider extends EventEmitter implemen
     const resourceUri = encodeUriComponentStrict(resourceString);
     this._sign(resourceUri, newExpiry, (err, signature) => {
       if (err) {
-        if (callback) {
-          callback(err);
-        } else {
-          this.emit('error', err);
-        }
+        callback(err);
       } else {
         this._currentTokenExpiryTimeInSeconds = newExpiry;
         this._credentials.sharedAccessSignature = signature;
 
-        const nextRenewalTimeout = (this._tokenValidTimeInSeconds - this._tokenRenewalMarginInSeconds) * 1000;
-        this._renewalTimeout = setTimeout(() => this._renewToken(), nextRenewalTimeout);
+        callback(null, this._credentials);
+      }
+    });
+  }
+
+  private _expiryTimerHandler(): void {
+    if (this._renewalTimeout) {
+      clearTimeout(this._renewalTimeout);
+    }
+
+    this._renewToken((err) => {
+      if (!err) {
+        this._scheduleNextExpiryTimeout();
         /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_005: [Every time a new token is created, the `newTokenAvailable` event shall be fired with the updated credentials.]*/
         this.emit('newTokenAvailable', this._credentials);
```

Comment:
>this.emit [](start = 8, length = 9)

maybe dangerous and unnecessary, but do you want to call emit inside setImmediate so the stack has been popped?

Reply:
I mean dangerous to change and maybe we don't need to change it.

---
In reply to: [197950701](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197950701) [](ancestors = 197950701)

Reply:
why would I pop the stack? isn't it easier to debug when we have the full stack?

---
In reply to: [197950859](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197950859) [](ancestors = 197950859,197950701)

Reply:
It is much easier to debug, but if this event handler makes it into user code in the current state, then our code is blocked on user code.



---
In reply to: [197951178](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197951178) [](ancestors = 197951178,197950859,197950701)

Reply:
aren't we the only users of this though?

---
In reply to: [197996374](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197996374) [](ancestors = 197996374,197951178,197950859,197950701)

Reply:
Good point.

(I'm not arguing to change this.  I'm just going through the mental exercise here.)

We're the only ones who use this event, but there's potential reentrancy issues here.  In device\transport\mqtt, the newTokenAvailable handler disconnects the transport.  I can't look in the debugger now, but it's possible that the base mqtt object is already disconnected by the time this call to emit() returns.  Now, here's where I'm waving my arms and being all theoretical, if the user happened to call Client.setOption in an event handler originally triggered by common/mqtt, then there's some event triggered by the base transport down deep in the stack, which ends up triggering a token renewal up here, which ends up triggering a disconnection higher in the stack in the middle of the emit call.  The base transport _should_ be hardened against stuff like this, but, well, you know how that goes.  

On the other hand, maybe I'm just confusing myself with too many different threading models in my head at the same time.  

The code was fine before without the setImmediate, so we shouldn't change it now.  

---
In reply to: [197997124](https://github.com/Azure/azure-iot-sdk-node/pull/330#discussion_r197997124) [](ancestors = 197997124,197996374,197951178,197950859,197950701)